snuba/datasets/plans/query_plan.py

```diff
@@ -17,32 +18,94 @@
     [Union[Query, CompositeQuery[Table]], RequestSettings, Reader], QueryResult
 ]
 
+TQuery = TypeVar("TQuery", bound=AbstractQuery)
+
 
 @dataclass(frozen=True)
-class ClickhouseQueryPlan:
+class QueryPlan(ABC, Generic[TQuery]):
     """
-    Provides the directions to execute a Clickhouse Query against one storage
-    or multiple joined ones.
-    This is produced by ClickhouseQueryPlanBuilder (provided by the dataset)
-    after the dataset query processing has been performed and the storage/s
-    has/have been selected.
-    It embeds the Clickhouse Query (the query to run on the storage after translation),
-    and the sequence of storage specific QueryProcessors to apply to the query after
-    the the storage has been selected. These have to be executed only once per plan
-    contrarily to the DB Query Processors, still provided by the storage, which
+    Provides the directions to execute a Clickhouse Query against one
+    storage or multiple joined ones.
+
+    This is produced by a QueryPlanner class (provided either by the
+    dataset or by an entity) after the dataset query processing has
+    been performed and the storage/s has/have been selected.
+
+    It embeds the Clickhouse Query (the query to run on the storage
+    after translation). It also provides a plan execution strategy
+    that takes care of coordinating the execution of the query against
+    the database. The execution strategy can also decide to split the
+    query into multiple chunks.
+
+    TODO: Bring the methods that apply the processors in the plan itself.
+    Now that we have two Plan implementations with a different
+    structure, all code depending on this would have to be written
+    differently depending on the implementation.
+    Having the methods inside the plan would make this simpler.
+    """
+
+    query: TQuery
+    execution_strategy: QueryPlanExecutionStrategy[TQuery]
+
+
+@dataclass(frozen=True)
+class ClickhouseQueryPlan(QueryPlan[Query]):
+    """
+    Query plan for a single entity, single storage query.
+
+    It provides the sequence of storage specific QueryProcessors
+    to apply to the query after the the storage has been selected.
+    These are divided in two sequences: plan processors and DB
+    processors.
+    Plan processors have to be executed only once per plan contrarily
+    to the DB Query Processors, still provided by the storage, which
     are executed by the execution strategy at every DB Query.
-    It also provides a plan execution strategy that takes care of coordinating the
-    execution of the query against the database. The execution strategy can also decide
-    to split the query into multiple chunks.
     """
 
-    query: Query
     plan_query_processors: Sequence[QueryProcessor]
```

Comment:
It seems like the SubqueryProcessors class could be used here too.

Reply:
Mind if I change this on top of #1510 ? Doing it here would be quite hard to merge as these are used a lot there.